pybmix/core/mixing.py

```diff
@@ -178,3 +178,140 @@ def _check_args(self, strength, discount):
         if discount >= 1 or discount <= 0:
             raise ValueError(
                 "Parameter 'discount' must be in the range (0, 1)")
+
+
+class StickBreakMixing(BaseMixing):
+    """ This class represents a Stick-Breaking process used for mixing in a mixture
+    model. A Stick-Breaking process with 'H' components is defined as follows:
+    
+        w[0] = v[0],
+        w[h] = v[h] x (1 - v[0])(1 - v[1])...(1 - v[h-1])   h=1, ... H-2
+        w[H-1] = 1 - sum(w[:H-1])
+
+    The stick proportions v are assumed to be independently distributed
+    v[h] ~ Beta(a_h, b_h). Different choices of a_h and b_h define different
+    processes. 
+    For instance a_h = 1, b_h = alpha defines a truncation of the
+    Dirichlet proces with total_mass = alpha. 
+    a_h = 1 - discount, b_h = strength + (h+1) * discount corresponds to a 
+    truncation of the Pitman-Yor process. 
+
+    Parameters
+    ----------
+    n_comp : int greater than 1
+        number of components in the process
+    strength : float greater than 0 or None
+        strength (or concentration) parameter of the Pitman-Yor Process
+    discount : float, in the range (0, 1) or None. 
+        If discount == 0, PitmanYorMixing is the same of DirichletProcessMixing
+    beta_params : sequence of tuples or None
+        The parameters of the Beta distributions of the stick proportions
+    """
+
+    ID = mixing_id.TruncSB
+    NAME = mixing_id.MixingId.Name(ID)
+
+    def __init__(self, n_comp, strength=None, discount=None, beta_params=None):
+        self._check_args(n_comp, strength, discount, beta_params)
+        self._build_prior_proto(n_comp, strength, discount, beta_params)
+
+    def prior_cluster_distribution(self, grid, nsamples, mc_iter=2000):
+        """
+        Evaluates the prior probability of the number of clusters on a
+        grid
+
+        Parameters
+        ----------
+        grid : array-like of shape (n,)
+               Points where to evaluate the probability mass function
+        nsamples : int
+                   Number of samples
+        mc_iter : int
+                 Number of Monte Carlo iterations used to approximate this
+                 distribution.
+        """
+
+        niter = int(mc_iter)
+        nus = self._simulate_weights(niter)
+        weights = self._break_sticks(nus)
+        # simulate from the categorical distribution in a smart way
+        cum_prob = np.cumsum(weights, axis=-1)
+        r = np.random.uniform(size=(nsamples, niter, 1))
+        clus_allocs = np.argmax(cum_prob > r, axis=-1)
+        mc_samples = np.apply_along_axis(lambda x: len(np.unique(x)), axis=0,
+                                   arr=clus_allocs)
+        out = np.zeros_like(grid, dtype=np.float)
```

Comment:
"in a smart way" indeed. As in too smart for me. How does this work exactly?
Also, do we need something like this in the C++ library?

Reply:
I updated the comment string, it's basically an inverse-cdf method. We don't need it in the C++ library since for loops are fast there, but sampling 10'000 categorical distributions in a for loop in python was too damn slow!